File: tests/fixtures/smelly_tests.rs
// テスト臭いのサンプル集 — savanna-smell-detector の動作確認用

#[cfg(test)]
mod tests {
    use std::thread;
    use std::time::Duration;

    // 1. Empty Test — テストが空っぽ
    #[test]
    fn test_empty() {
    }

    // 2. Missing Assertion — アサーションなし
    #[test]
    fn test_no_assertion() {
        let x = 1 + 1;
        let _y = x * 2;
    }

    // 3. Sleepy Test — sleep を使っている
    #[test]
    fn test_sleepy() {
        thread::sleep(Duration::from_secs(2));
        assert_eq!(1, 1);
    }

    // 4. Conditional Test Logic — テスト内に条件分岐
    #[test]
    fn test_conditional() {
        let x = 42;
        if x > 0 {
            assert!(true);
        } else {
            assert!(false);
        }
    }

    // 5. Ignored Test — 無視されたテスト
    #[test]
    #[ignore]
    fn test_ignored() {
        assert_eq!(1, 1);
    }

    // 6. Redundant Print — printlnが残っている
    #[test]
    fn test_with_print() {
        println!("debug: calculating...");
        let result = 2 + 2;
        dbg!(result);
        assert_eq!(result, 4);
    }

    // 7. Assertion Roulette — メッセージなしの複数アサーション
    #[test]
    fn test_assertion_roulette() {
        let x = 42;
        assert!(x > 0);
        assert_eq!(x, 42);
        assert_ne!(x, 0);
    }

    // 8. Magic Number Test — マジックナンバー
    #[test]
    fn test_magic_number() {
        let result = compute();
        assert_eq!(result, 86400);
        assert_eq!(result % 3600, 0);
    }

    fn compute() -> i64 { 86400 }

    // 9. 良いテスト — 臭いなし
    #[test]
    fn test_clean() {
        let result = 2 + 2;
        assert_eq!(result, 4);
    }

    // 10. Ignored Test with reason — 理由付きは検出しない
    #[test]
    #[ignore = "GUI environment required"]
    fn test_ignored_with_reason() {
        assert_eq!(1, 1);
    }

    // 11. Table-driven test — テーブル駆動テストは conditional ではない
    #[test]
    fn test_table_driven() {
        let cases: &[(&str, i32)] = &[
            ("one", 1),
            ("two", 2),
            ("three", 3),
        ];
        for (input, expected) in cases {
            assert_eq!(parse_num(input), *expected, "failed for: {input}");
        }
    }

    fn parse_num(_s: &str) -> i32 { 1 }
}

// #[test]
// fn test_commented_out() {
//     assert_eq!(1, 1);
// }

File: tests/fixtures/smelly_tests.py
# テスト臭いのサンプル集 — pytest / unittest 形式
import time
import unittest

import pytest


# 1. Empty Test
def test_empty():
    pass


# 2. Missing Assertion
def test_no_assertion():
    result = 2 + 2
    print(result)


# 3. Sleepy Test
def test_sleepy():
    time.sleep(1)
    assert True


# 4. Conditional Test Logic
def test_conditional():
    if some_condition():
        assert value() == 1
    else:
        assert value() == 2


# 5. Ignored Test (デコレーター)
@pytest.mark.skip(reason="not ready")
def test_skipped():
    assert True


# 6. Redundant Print
def test_with_print():
    result = compute()
    print("result:", result)
    assert result == 4


# 7. Assertion Roulette（メッセージなしの複数アサーション）
def test_assertion_roulette():
    assert first() == 1
    assert second() == 2
    assert third() == 3


# 8. Magic Number Test
def test_magic_number():
    assert calculate_total() == 12345


# 9. Silent Skip（early return）
def test_early_return():
    if not server_available():
        return
    assert ping() == "pong"


# 10. Fragile Test（sleep + 時間依存の成功判定）
def test_fragile_timing():
    start = time.time()
    time.sleep(1)
    run_task()
    assert time.time() - start < 5


# 11. Commented-Out Test
# def test_commented_out():
#     assert 1 == 1


# 12. Clean test（スメルなし）
def test_clean():
    result = 2 + 2
    assert result == 4, "2 + 2 should equal 4"


# unittest 形式
class TestUnittestStyle(unittest.TestCase):
    def test_empty_method(self):
        pass

    def test_unittest_assertions(self):
        self.assertEqual(2 + 2, 4, "addition should work")

    @unittest.skip("legacy")
    def test_skipped_method(self):
        self.assertTrue(True)

    def test_skip_in_body(self):
        self.skipTest("runtime skip")

    def test_sleep_in_loop(self):
        for i in range(3):
            time.sleep(1)
            self.assertLess(i, 3)


# pytest.raises は例外アサーションとして扱う
def test_raises():
    with pytest.raises(ValueError):
        int("not a number")


# async テスト
async def test_async_sleepy():
    import asyncio
    await asyncio.sleep(1)
    assert True
